Adding Snapshot Options to deal with Server Timestamps

schmidt-sebastian opened the pull request:
This adds to new deserialization options for pending server timestamps - returning the previous value or returning an estimated timestamp.

Port of https://github.com/firebase/firebase-ios-sdk/pull/537

mikelehen commented on packages/firestore/test/integration/api/server_timestamp.test.ts:
```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          verifyTimestampsAreResolved(snapshot);
+        })
```
Lines 245-251 are repeated from the previous test, right?  Why?  It's really hard to read tests that are long and repetitive. I can't figure out what's setup vs behavior being tested (comments can be used as a crutch, but you didn't do that either :-)).

My preferred way to do this test (in the spirit of DRY) would probably be to change writeInitialData() to accept an optional initial value for the when fields.

Then you can just do writeInitialData(/*whenValue=*/42).then(() => docRef.update(updateData)) and then wait for the events / do your validation.

schmidt-sebastian replied:
I cleaned up the tests quite a bit and removed the parts that are exercised in other tests. I think this reduces the scope and duplication enough to not introduce further helpers.

mikelehen commented on packages/firestore/test/integration/api/server_timestamp.test.ts:
```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
+          );
+          expect(
+            snapshot.get('a', { serverTimestamps: 'estimate' })
+          ).to.be.instanceof(Date);
```
Are the none / estimate tests actually valuable?  We already test their behavior elsewhere. Else, please update the name of the test.

mikelehen commented on packages/firestore/test/integration/api/server_timestamp.test.ts:
```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
+          ).to.be.instanceof(Date);
+          expect(
+            snapshot.get('a', { serverTimestamps: 'estimate' })
+          ).to.be.instanceof(Date);
```
Again, I'd prefer not throw in extra expectations just because it's convenient.  Testing the behavior of 'estimate' and 'previous' on actual resolved timestamps (aka normal Dates) is probably good, but should go elsewhere (either in the existing 'can return {previous,estimate}' tests or in a new test.

schmidt-sebastian replied:
Removed

mikelehen commented on packages/firestore/test/integration/api/server_timestamp.test.ts:
```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
+        .then(snapshot => {
+          verifyTimestampsUsePreviousValue(snapshot, null);
+          return waitForRemoteEvent();
+        })
```
Can you see if you can clean up the setup here too?

schmidt-sebastian replied:
Done

mikelehen commented on packages/firestore/test/integration/api/server_timestamp.test.ts:
```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
+    });
+  });
+
+  it('uses previous value from local mutation', () => {
```
It took me a couple minutes to figure out what was different about this test vs the previous one.  These tests really need less boilerplate and/or more comments. Or perhaps better would actually be to move these to unit tests rather than integration tests so you can more directly test the applyToLocalView() code without all of this extra setup / asynchrony / etc. in the way... :-/  The integration tests are for hitting all of the public surface area and sanity-checking behavior rather than covering all edge cases.

schmidt-sebastian replied:
I took a quick look at testing this via the mutations batches earlier, and some of this testing can indeed be moved there by verifying `previousState` via `valueWithOptions`.
Since Server Timestamps were already mostly tested via these integration tests, I ended up shying away from this though. 
Before I start reworking this, I would like to know how strong you feel about it, as this change should then also be ported to the Android and iOS clients.

schmidt-sebastian commented on packages/firestore/src/api/database.ts:
```diff
@@ -982,6 +984,9 @@ export class DocumentReference implements firestore.DocumentReference {
 }
 
+/** Options interface that can be provided to configure the deserialization of DocumentSnapshots. */
+export interface SnapshotOptions extends firestore.SnapshotOptions {}
```
We don't do this with the other options. The other options only have boolean values and we just access them directly (options.merge for example). I added this so I don't have to import the public type in field_value.ts.

schmidt-sebastian commented on packages/firestore/src/api/database.ts:
```diff
@@ -1702,6 +1723,25 @@ function validateSetOptions(
 
+function validateSnapshotOptions(
+  methodName: string,
+  options?: firestore.SnapshotOptions
```
That makes sense. Thanks for explaining the reasoning behind the validateSetOptions() signature.

schmidt-sebastian commented on packages/firestore/src/model/field_value.ts:
```diff
@@ -61,6 +62,34 @@ export enum TypeOrder {
+      case 'previous':
+        return new FieldValueOptions(ServerTimestampBehavior.Previous);
+      case 'none': // Fall-through intended.
+      default:
```
That works.

schmidt-sebastian commented on packages/firestore/src/model/field_value.ts:
```diff
@@ -80,7 +109,7 @@ export type FieldType = null | boolean | number | string | {};
   readonly typeOrder: TypeOrder;
 
-  abstract value(): FieldType;
+  abstract value(options?: FieldValueOptions): FieldType | undefined;
```
It looks like it is a mistake. ServerTimestamps return null but never undefined.

schmidt-sebastian commented on packages/firestore/src/model/field_value.ts:
```diff
@@ -319,18 +348,31 @@ export class TimestampValue extends FieldValue {
-  constructor(readonly localWriteTime: Timestamp) {
+  constructor(
+    readonly localWriteTime: Timestamp,
+    readonly previousValue: FieldValue | null
```
Gil and I talked about overlays when we discussed this and hope that overlays make getting the previous value a lot easier.

schmidt-sebastian commented on packages/firestore/src/model/field_value.ts:
```diff
@@ -319,18 +348,31 @@ export class TimestampValue extends FieldValue {
+        case ServerTimestampBehavior.Estimate:
+          return this.localWriteTime.toDate();
+        case ServerTimestampBehavior.Previous:
+          return this.previousValue ? this.previousValue.value(options) : null;
```
It falls through to the function's main return value.

I changed this to an if/else to make this more obvious (and to not repeat the same error handling everywhere).

schmidt-sebastian commented on packages/firestore/src/model/mutation.ts:
```diff
@@ -226,13 +226,17 @@ export abstract class Mutation {
    *     the client has no knowledge of the pre-mutation state of the document.
+   * @param baseDoc The state of the document prior to this mutation batch. The
+   *     input document can be null if the client has no knowledge of the
+   *     pre-mutation state of the document.
```
`set({a:FieldValue.serverTimestamp()})` is a good example why this doesn't work. It will create two mutations: One that clears the map `a` and removes the existing value, and one that adds the timestamp. When we apply the timestamp, `previousValue` has already been cleared in `maybeDoc`.

schmidt-sebastian commented on packages/firestore/src/model/mutation.ts:
```diff
@@ -539,14 +549,27 @@ export class TransformMutation extends Mutation {
+
+        if (baseDoc instanceof Document) {
+          previousValue =
+            (baseDoc as Document).field(fieldTransform.field) || null;
```
Oh, yes. I remember this now from a TS talk I watched I while ago :)

schmidt-sebastian commented on packages/firestore/src/typings/firestore.d.ts:
```diff
@@ -518,6 +518,29 @@ declare namespace firestore {
+     * been set to their final value.
+     *
+     * By specifying `estimate`, pending server timestamps return a local
+     * estimate. This estimate will likely differ from the final value and may
```
Done.

schmidt-sebastian commented on packages/firestore/src/typings/firestore.d.ts:
```diff
@@ -518,6 +518,29 @@ declare namespace firestore {
+     *
+     * By specifying `estimate`, pending server timestamps return a local
+     * estimate. This estimate will likely differ from the final value and may
+     * cause these values to change once the server result becomes available.
```
Reworded.

schmidt-sebastian commented on packages/firestore/src/typings/firestore.d.ts:
```diff
@@ -518,6 +518,29 @@ declare namespace firestore {
+     * estimate. This estimate will likely differ from the final value and may
+     * cause these values to change once the server result becomes available.
+     *
+     * By specifying `previous`, these pending timestamps will instead retain
```
Removed

schmidt-sebastian commented on packages/firestore/src/typings/firestore.d.ts:
```diff
@@ -518,6 +518,29 @@ declare namespace firestore {
+     * cause these values to change once the server result becomes available.
+     *
+     * By specifying `previous`, these pending timestamps will instead retain
+     * their previous value.
```
Reworded.

schmidt-sebastian commented on packages/firestore/src/typings/firestore.d.ts:
```diff
@@ -518,6 +518,29 @@ declare namespace firestore {
+     * By specifying `previous`, these pending timestamps will instead retain
+     * their previous value.
+     *
+     * If omitted or set to `none`, `null` will be returned by default until the
```
Fixed here and above.

schmidt-sebastian commented on packages/firestore/src/util/input_validation.ts:
```diff
@@ -169,6 +169,58 @@ export function validateNamedOptionalType(
+ */
+export function validateNamedPropertyEquals<T>(
+  functionName: string,
+  inputName: T,
```
Fixed.

schmidt-sebastian commented on packages/firestore/src/util/input_validation.ts:
```diff
@@ -169,6 +169,58 @@ export function validateNamedOptionalType(
+
+  for (let val of expected) {
+    if (val === input) {
+      return true;
```
As written, it returns boolean. Changed to return void.

schmidt-sebastian commented on packages/firestore/test/integration/api/server_timestamp.test.ts:
```diff
@@ -63,35 +63,72 @@ apiDescribe('Server Timestamps', persistence => {
-      expect(remoteSnap.data()).to.deep.equal(expectedDataWithTimestamp(when));
-    });
+    }
   }
```
GIl asked me to split up the await logic from the verification. Unfortunately, this blows up the JS tests quite a bit. I'll keep the suggested cleanup in mind for when I need to ask for a favor :)

schmidt-sebastian commented on packages/firestore/test/integration/api/validation.test.ts:
```diff
@@ -241,6 +241,28 @@ apiDescribe('Validation:', persistence => {
+        expect(() =>
+          snapshot.data({ serverTimestamps: 'foo' } as any)
+        ).to.throw(
+          `The serverTimestamps property of the options argument for DocumentSnapshot.data() is expected to be one of "estimate","previous","none",  but was "foo".`
```
Not using a space was my somewhat questionable workaround for not adding the "or" (I originally had a regexp-like thing in the error message that used `|` as a separator). Putting the values at the end nicely works around this problem. Thanks!

schmidt-sebastian commented:
I addressed all comments. I also pretended that Greg's disableNetwork PR already landed as this allowed me to clean up the tests further.

If you have time, I'd appreciate another round on this. Otherwise, we can wait until his PR is merged and then proceed with this review.

schmidt-sebastian commented on packages/firestore-types/index.d.ts:
```diff
@@ -565,20 +588,34 @@ export class DocumentSnapshot {
    * document or field doesn't exist.
    *
+   * By default, a `FieldValue.serverTimestamp()` that have not yet been set to
+   * its final value will be returned as `null`. You can override this by
```
Fixed.

schmidt-sebastian commented on packages/firestore/src/model/field_value.ts:
```diff
@@ -349,7 +351,7 @@ export class TimestampValue extends FieldValue {
  *   serialized.
  * - When evaluated locally (e.g. for snapshot.data()), they by default
- *   evaluate to `null`. This behavior can be configured by passing custom
+ *   evaluate to 'null'. This behavior can be configured by passing custom
```
Reverted.

schmidt-sebastian commented on packages/firestore/test/integration/api/validation.test.ts:
```diff
@@ -258,7 +258,7 @@ apiDescribe('Validation:', persistence => {
           snapshot.data({ serverTimestamps: 'foo' } as any)
         ).to.throw(
-          `The serverTimestamps property of the options argument for DocumentSnapshot.data() is expected to be one of "estimate","previous","none",  but was "foo".`
+          `Invalid value "foo" provided to function DocumentSnapshot.data() for option 'serverTimestamps'. Acceptable values: estimate, previous, none`
```
I like the single quotes more. The double quotes come from JSON.stringify. In the interest of consistency, I persuaded myself to use double quotes everywhere.
